From 87f9a90784c10aa4dd07869380b85596ff20976d Mon Sep 17 00:00:00 2001
From: William Bright <William.Bright@imd-tec.com>
Date: Wed, 15 Apr 2026 11:57:54 +0100
Subject: [PATCH] arm64: dts: qcom: qcs8550-imdt-sbc: Add PCIe switch Key B
 overlay

Add a device tree overlay to select M.2 Key B PCIe routing instead of
the default GBE PHY routing.

The PCIe switch SEL line (TL37 gpio16) selects GBE PHY routing when
low and M.2 Key B routing when high. The base board dts drives it low
via the pcie_switch_sel_default pinctrl state, applied from pcie1's
pinctrl-0 at controller probe, immediately before link training.

The overlay therefore replaces pcie1's pinctrl-0 with a state that
drives the pin high instead. The property is replaced wholesale
because a devicetree overlay cannot delete the output-low property
from the existing pcie_switch_sel_default node; and a gpio-hog is
deliberately not used, as any hog setting would be overridden by the
pinctrl state pcie1 applies later in boot.

Signed-off-by: William Bright <william.bright@imd-tec.com>
---
 arch/arm64/boot/dts/qcom/Makefile             |  1 +
 .../dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso  | 36 ++++++++++++++++++
 2 files changed, 37 insertions(+)
 create mode 100644 arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso

diff --git a/arch/arm64/boot/dts/qcom/Makefile b/arch/arm64/boot/dts/qcom/Makefile
index 79fd05a32e00..a9a57392b4c9 100644
--- a/arch/arm64/boot/dts/qcom/Makefile
+++ b/arch/arm64/boot/dts/qcom/Makefile
@@ -221,6 +221,7 @@ dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-aim300-aiot.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-imdt-sbc.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-imdt-sbc-display.dtbo
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-imdt-sbc-ar1335-csi0.dtbo
+dtb-$(CONFIG_ARCH_QCOM)	+= qcs8550-imdt-sbc-pcie-keyb.dtbo
 
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs9100-ride.dtb
 dtb-$(CONFIG_ARCH_QCOM)	+= qcs9100-ride-r3.dtb
diff --git a/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso b/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso
new file mode 100644
index 000000000000..3c1a5c9b7d42
--- /dev/null
+++ b/arch/arm64/boot/dts/qcom/qcs8550-imdt-sbc-pcie-keyb.dtso
@@ -0,0 +1,36 @@
+// SPDX-License-Identifier: BSD-3-Clause
+/*
+ * Copyright (c) 2026 IMD Technologies Ltd
+ *
+ * There are two options for PCIe routing on the IMDT QCS8550 SBC.
+ * The PCIe switch SEL line (TLMM gpio16) selects GBE PHY routing when
+ * low (the default, driven by the pcie_switch_sel_default pinctrl
+ * state applied from pcie1's pinctrl-0) and M.2 Key B routing when
+ * high.
+ *
+ * This overlay reroutes PCIe1 to the M.2 Key B slot by replacing the
+ * SEL pinctrl state with one that drives the pin high. pcie1's
+ * pinctrl-0 is replaced wholesale because a devicetree overlay cannot
+ * delete the output-low property from the existing
+ * pcie_switch_sel_default state node; a gpio-hog is deliberately not
+ * used, as any hog setting is overridden by the pinctrl state pcie1
+ * applies later in boot, immediately before link training.
+ */
+/dts-v1/;
+/plugin/;
+
+&tlmm {
+	pcie_switch_sel_keyb: pcie-switch-sel-keyb-state {
+		pins = "gpio16";
+		function = "gpio";
+		drive-strength = <2>;
+		bias-disable;
+		output-high;
+	};
+};
+
+&pcie1 {
+	pinctrl-0 = <&pcie1_default_state>,
+		    <&pcie_switch_sel_keyb>,
+		    <&gbe_reset_default>;
+};
-- 
2.43.0
